c/examples/cpp_sorting_example.cpp

```diff
@@ -0,0 +1,121 @@
+#include <cstddef>
+#include <vector>
+#include <algorithm>
+#include <stdexcept>
+#include <sstream>
+#include <iostream>
+#include <type_traits>
+#include <tskit.h>
+
+static void
+handle_tskit_return_code(int code)
+{
+    if (code != 0) {
+        std::ostringstream o;
+        o << tsk_strerror(code);
+        throw std::runtime_error(o.str());
+    }
+}
+
+struct edge_plus_time {
+    double time;
+    tsk_id_t parent, child;
+    double left, right;
+};
+
+int
+sort_edges(tsk_table_sorter_t *sorter, tsk_size_t start)
+{
+    // if (sorter->tables->edges.metadata != nullptr) {
+    //    throw std::invalid_argument(
+    //        "the sorter does not currently handle edge metadata");
+    //}
+    if (start != 0) {
+        throw std::invalid_argument("the sorter requires start==0");
+    }
+
+    std::vector<edge_plus_time> temp;
+    temp.reserve(static_cast<std::size_t>(sorter->tables->edges.num_rows));
+
+    const auto *edges = &sorter->tables->edges;
+    const auto *nodes = &sorter->tables->nodes;
+
+    // NOTE: this future-proofs the type against changing to int64_t
+    // for tsk_id_t.  It will also look odd to a C programmer,
+    // but the edges->parent[0] need not be valid at run time.
+    // Rather, it is checking what the return type of the expression
+    // would be, which is tsk_id_t& (with the & being due to decltype
+    // not returning the decay type).
+    const auto num_nodes
+        = static_cast<std::remove_reference<decltype(edges->parent[0])>::type>(
+            nodes->num_rows);
+
+    for (decltype(sorter->tables->edges.num_rows) i = 0;
+         i < sorter->tables->edges.num_rows; ++i) {
+        if (edges->parent[i] >= num_nodes || edges->child[i] >= num_nodes) {
```

Comment:
We don't have to worry about his within the sorter routine, as this will have been checked already by ``tsk_table_collection_check_integrity`` (unless we call sorter_run() with TSK_NO_CHECK_INTEGRITY)

Reply:
Yeah, so that's a little tricky.  I'm removing the check, but invalid ids can then fail silently.

Reply:
It can't though - tsk_table_collection_check_integrity has already checked this during ``sorter_init``, unless you explicitly tell it not to.